iroha/src/sumeragi/view_change.rs

```diff
@@ -0,0 +1,325 @@
+    payload: ProofPayload,
+    signatures: Signatures,
+}
+
+impl Proof {
+    /// Constructor for `CommitTimeout` view change suggestion.
+    pub fn commit_timeout(
+        voting_block_hash: Hash,
+        previous_proof: Hash,
+        latest_block: Hash,
+    ) -> Self {
+        Self {
+            signatures: Signatures::default(),
+            payload: ProofPayload {
+                reason: Reason::CommitTimeout(CommitTimeout { voting_block_hash }),
+                previous_proof,
+                latest_block,
+            },
+        }
+    }
+
+    /// Constructor for `BlockCreationTimeout` view change suggestion.
+    pub fn block_creation_timeout(
+        transaction_receipt: TransactionReceipt,
+        previous_proof: Hash,
+        latest_block: Hash,
+    ) -> Self {
+        Self {
+            signatures: Signatures::default(),
+            payload: ProofPayload {
+                reason: Reason::BlockCreationTimeout(BlockCreationTimeout {
+                    transaction_receipt,
+                }),
+                previous_proof,
+                latest_block,
+            },
+        }
+    }
+
+    /// Constructor for `NoTransactionReceiptReceived` view change suggestion.
+    pub fn no_transaction_receipt_received(
+        transaction_hash: Hash,
+        previous_proof: Hash,
+        latest_block: Hash,
+    ) -> Self {
+        Self {
+            signatures: Signatures::default(),
+            payload: ProofPayload {
+                reason: Reason::NoTransactionReceiptReceived(NoTransactionReceiptReceived {
+                    transaction_hash,
+                }),
+                previous_proof,
+                latest_block,
+            },
+        }
+    }
+
+    /// Hash of the proof `payload`.
+    pub fn hash(&self) -> Hash {
+        let bytes: Vec<u8> = self.payload.clone().into();
+        Hash::new(&bytes)
+    }
+    /// Signs this message with the peer's public and private key.
+    /// This way peers vote for changing the view - changing the roles of peers.
+    ///
+    /// # Errors
+    /// Can fail during creation of signature
+    pub fn sign(mut self, key_pair: &KeyPair) -> Result<Proof> {
+        let signature = Signature::new(key_pair.clone(), self.hash().as_ref())?;
+        self.signatures.add(signature);
+        Ok(self)
+    }
+
+    /// Verify if the proof is valid, given the peers in `topology`.
+    pub fn verify(&self, peers: &HashSet<PeerId>, max_faults: u32) -> bool {
+        let peer_public_keys: HashSet<PublicKey> = peers
+            .iter()
+            .map(|peer_id| peer_id.public_key.clone())
+            .collect();
+        let n_signatures = self
+            .signatures
+            .verified(self.hash().as_ref())
+            .iter()
+            .filter(|signature| peer_public_keys.contains(&signature.public_key))
+            .count();
+        #[allow(clippy::int_plus_one)]
+        {
+            n_signatures >= max_faults as usize + 1
+        }
```

Comment:
Why do we need to discard this warning though?
It is legit, you could write it like this:
```rs
n_signatures > max_faults as usize
```

Reply:
For better understand-ability. f+1 and 2f+1 are the numbers that are defined in the whitepaper, and it is easier if we keep them like that.

Reply:
I guess it should reference whitepaper then. Otherwise the reason is not clear